File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    void inorder(TreeNode* node, int& count, int k, int& ans){
        if(!node) return;
        inorder(node->left, count, k, ans);
        count++;
        if(count == k){
            ans = node->val;
            return;
        }
        inorder(node->right, count, k, ans);
    }

    int kthSmallest(TreeNode* root, int k) {
        int count = 0;
        int ans = 0;
        inorder(root, count, k, ans);
        return ans;
    }
};

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        def dfs(node, maximum):
            if not node:
                return 0

            if node.val >= maximum:
                total = 1
            else:
                total = 0
            
            total += dfs(node.left, max(node.val, maximum))
            total += dfs(node.right, max(node.val, maximum))

            return total
        
        return dfs(root, root.val)


        
        
